From 3c512efd635d6ea37b043850ed6411e0472c3d8a Mon Sep 17 00:00:00 2001
From: Min He <min.he@intel.com>
Date: Thu, 11 Jan 2018 13:06:41 +0800
Subject: [PATCH 0440/1060] vhm: fix issue in do_map_guest_phys

In do_map_guest_phys(), we need to add a correct offset to get
physical address of gpa. This patch can avoid kernel panic issue
when there're multiple mem_segs for a guest.

Change-Id: Iacef2c562ea7e0ec0b0cda02684633279ccefd36
Signed-off-by: Min He <min.he@intel.com>
Reviewed-by: Chi, Mingqiang <mingqiang.chi@intel.com>
Reviewed-by: Dong, Eddie <eddie.dong@intel.com>
Tested-by: Dong, Eddie <eddie.dong@intel.com>
---
 drivers/vhm/vhm_mm.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/drivers/vhm/vhm_mm.c b/drivers/vhm/vhm_mm.c
index deba62a2c02b..b9ba1a7baad3 100644
--- a/drivers/vhm/vhm_mm.c
+++ b/drivers/vhm/vhm_mm.c
@@ -388,7 +388,7 @@ static void *do_map_guest_phys(struct vhm_vm *vm, u64 guest_phys, size_t size)
 		}
 
 		mutex_unlock(&vm->seg_lock);
-		return phys_to_virt(seg->base + guest_phys);
+		return phys_to_virt(seg->base + guest_phys - seg->gpa);
 	}
 	mutex_unlock(&vm->seg_lock);
 	return NULL;
-- 
2.17.1

